x/crosschain/keeper/msg_server_vote_inbound_tx.go

```diff
@@ -85,127 +85,161 @@ func (k msgServer) VoteOnObservedInboundTx(goCtx context.Context, msg *types.Msg
+	cctx := types.InitializeCCTX(ctx, *msg, tss.TssPubkey)
 
-	// create the CCTX
-	cctx := k.CreateNewCCTX(
-		ctx,
-		msg,
-		index,
-		tssPub,
-		types.CctxStatus_PendingInbound,
-		msg.SenderChainId,
-		msg.ReceiverChain,
-	)
+	// 2. Validate the CCTX
+	err = cctx.Validate()
+	if err != nil {
+		return nil, err
+	}
 
-	defer func() {
-		EmitEventInboundFinalized(ctx, &cctx)
-		k.AddFinalizedInbound(ctx, msg.InTxHash, msg.SenderChainId, msg.EventIndex)
-		// #nosec G701 always positive
-		cctx.InboundTxParams.InboundTxFinalizedZetaHeight = uint64(ctx.BlockHeight())
-		cctx.InboundTxParams.TxFinalizationStatus = types.TxFinalizationStatus_Executed
-		k.RemoveInTxTrackerIfExists(ctx, cctx.InboundTxParams.SenderChainId, cctx.InboundTxParams.InboundTxObservedHash)
-		k.SetCctxAndNonceToCctxAndInTxHashToCctx(ctx, cctx)
-	}()
+	// 3. Process the inbound CCTX, the process function manages the state commit and cctx status change.If the process fails the changes to the evm state are rolled back.
+	k.ProcessInbound(ctx, &cctx)
+	// 4. Save the inbound CCTX to the store.This is called irrespective of the status of the CCTX or the outcome of the process function.
+	k.SaveInbound(ctx, &cctx, msg.EventIndex)
+	return &types.MsgVoteOnObservedInboundTxResponse{}, nil
+}
 
-	// FinalizeInbound updates CCTX Prices and Nonce
-	// Aborts is any of the updates fail
-	if common.IsZetaChain(msg.ReceiverChain) {
-		tmpCtx, commit := ctx.CacheContext()
-		isContractReverted, err := k.HandleEVMDeposit(tmpCtx, &cctx, *msg, msg.SenderChainId)
-
-		if err != nil && !isContractReverted { // exceptional case; internal error; should abort CCTX
-			cctx.CctxStatus.ChangeStatus(types.CctxStatus_Aborted, err.Error())
-			return &types.MsgVoteOnObservedInboundTxResponse{}, nil
-		} else if err != nil && isContractReverted { // contract call reverted; should refund
-			revertMessage := err.Error()
-			chain := k.zetaObserverKeeper.GetSupportedChainFromChainID(ctx, cctx.InboundTxParams.SenderChainId)
-			if chain == nil {
-				cctx.CctxStatus.ChangeStatus(types.CctxStatus_Aborted, "invalid sender chain")
-				return &types.MsgVoteOnObservedInboundTxResponse{}, nil
-			}
+// InitializeCCTX returns a new CCTX from a given inbound message.
 
-			gasLimit, err := k.GetRevertGasLimit(ctx, cctx)
-			if err != nil {
-				cctx.CctxStatus.ChangeStatus(types.CctxStatus_Aborted, "can't get revert tx gas limit"+err.Error())
-				return &types.MsgVoteOnObservedInboundTxResponse{}, nil
-			}
-			if gasLimit == 0 {
-				// use same gas limit of outbound as a fallback -- should not happen
-				gasLimit = msg.GasLimit
-			}
+// ProcessInbound processes the inbound CCTX.
+// It does a conditional dispatch to ProcessZEVMDeposit or ProcessCrosschainMsgPassing based on the receiver chain.
+// The internal functions handle the state changes and error handling.
+func (k Keeper) ProcessInbound(ctx sdk.Context, cctx *types.CrossChainTx) {
```

Comment:
Keeping it in this might be a good option, this function is only called at one place which is the Inbound Vote.

Reply:
Yeah but this would change in the future:
- We work on IBC integration where inbound are automatically processed from IBC messages without the need of this message
- We integrate the header/proof system, there would be therefore a separate message that would still process the inbound without the need to vote on it.
Make this separate when the processing is independing from this message clear now would make it easier adding these new functionalities in the future

Reply:
2c856b06d8026602ae5118c27cf657bf42a04bb9